Catnip docs (#353)
configcat/docs: website/docs/targeting/setting-evaluation.md

---
id: setting-evaluation
title: Setting Evaluation
description: This document offers an in-depth guide on how the SDK determines the value of a setting.
---

# Setting Evaluation

This document offers an in-depth guide on how the SDK determines the value of a setting when the `GetValue` call is made. Understanding this process requires prior knowledge of targeting concepts.

The setting's value is influenced by:

> - Rules established on the Dashboard,
> - The [User Object] provided to the `GetValue` function, and
> - The `default value` assigned in the `GetValue` function.

Reviewer: ```suggestion
- The feature flag's rules defined on the Dashboard,
- The [User Object] provided to the `GetValue` function, and
- The `default value` passed to the `GetValue` function.

- Rules established on the Dashboard,
- The `default value` assigned in the `GetValue` function.
```

### Evaluation of a Prerequisite Flag Condition

The SDK evaluates the prerequisite flag as a separate setting using the same [User Object]. The result of this evaluation is compared with the Dashboard's set value using the specified comparator, resulting in a `true` or `false` outcome.

If an error occurs during this evaluation, the process stops, and the SDK returns the default value.

### Evaluation of a Segment Condition

The SDK evaluates segment conditions similarly to [user conditions].

If the user condition is `true` and the comparator is `IS IN SEGMENT`, or if the user condition is `false` and the comparator is `IS NOT IN SEGMENT`, the segment condition yields `true`.

Conversely, if the user condition is `false` and the comparator is `IS IN SEGMENT`, or if the user condition is `true` and the comparator is `IS NOT IN SEGMENT`, the segment condition yields `false`.

If the user condition is `indeterminable (cannot evaluate)`, so is the segment condition.

## Evaluation of Percentage Options

Percentage options are designed to be consistent and sticky across all SDKs, ensuring a reliable experience.

> This approach primarily relies by default on the identifier ([this can be changed to any attribute]) from the [User Object] provided to the SDK's `getValue()` function. Here's how it works:

Reviewer: ```suggestion
By default, this feature relies on the `Identifier` attribute of the [User Object]. (The attribute to use as the basis of the grouping [can be changed to any other attribute] though.) Here's how it works:
```

> - The SDKs create a hash from the combination of the [User Object]'s attribute and the specific feature flag's `Key`.
> - This hash process assigns a unique 0-99 number to each User for a specific feature flag.
> - The assigned number determines the user's eligibility for a feature flag based on the percentage options set on the Dashboard.

Reviewer: ```suggestion
- The SDKs create a hash from the combination of the value of the percentage evaluation attribute (the [User Object] attribute to use as the basis of the grouping) and the specific feature flag's name.
- The hash algorithm assigns a number between 0 and 99 to the user. The number will be the same for users who have the same percentage evaluation attribute.
- The assigned number determines which group the user falls into, i.e. which percentage option applies to the user. 
```

- Importantly, this number remains fixed and consistent for each User across all SDKs, ensuring uniformity in feature flag evaluation.

:::caution

> By hashing both the User's attribute and the feature flag's key, we ensure diverse user bases for different feature flags. This method prevents the same group of users from being repeatedly targeted across various feature flags.

Reviewer: ```suggestion
By hashing the combination of the user attribute and the feature flag's name, we ensure diverse user groups for different feature flags. In other words, this method prevents the same group of users from being repeatedly targeted across various feature flags.
```

:::

:::info

> The evaluation process is entirely contained within the SDKs, meaning your User's sensitive data never leaves your system. The data flow is one-way – from ConfigCat CDN servers to your SDKs – and ConfigCat does not receive or store any attributes of the [User Object] passed to the SDKs. This design prioritizes the privacy and security of user data.

Reviewer: ```suggestion
The evaluation process is entirely implemented within the SDKs, meaning your users' sensitive data never leaves your system. The data flow is one-way – from ConfigCat CDN servers to your SDKs – and ConfigCat does not receive or store any attributes of the [User Object] passed to the SDKs. This design prioritizes the privacy and security of user data.
```

:::

### Percentage Evaluation Attribute

> The default attribute used for percentage evaluation is the [User Object]'s `Identifier`. However, you can change this to any attribute you want.

Reviewer: ```suggestion
By default the user attribute used for percentage evaluation is `Identifier`. However, you can change this to any other attribute you want.
```

*TODO - Where and how to change this*

### Example Scenarios for Percentage-Based Targeting

> Imagine you have two users, Jane and Joe, and you're experimenting with two different feature flags (`isTwitterSharingEnabled` and `isFacebookSharingEnabled`) that use percentage-based targeting. In these scenarios, we see how percentage-based targeting allows for a controlled and gradual rollout of features, ensuring a smooth transition for users like Jane and Joe.

Reviewer: ```suggestion
Imagine you have two users, Jane and Joe, and you're experimenting with two different feature flags (`isTwitterSharingEnabled` and `isFacebookSharingEnabled`) that use percentage-based targeting. In these scenarios, we see how percentage options allow for a controlled and gradual rollout of features, ensuring a smooth transition for users like Jane and Joe.
```

> First, the users are assigned a number between 0-99 based on the hash of their identifier and the feature flag's key. This number determines their eligibility for a feature flag based on the percentage options set on the Dashboard.

Reviewer: ```suggestion
First, the users are assigned a number between 0-99 based on the hash of their identifier and the feature flag's name. This number determines their eligibility for a feature based on the percentage options set on the Dashboard.
```

| Joe  | `hash('Joe' + 'isTwitterSharingEnabled') mod 100` <br/> Results in **32** | `hash('Joe' + 'isFacebookSharingEnabled') mod 100` <br/> Results in **12**  |

1. **Initial Setting: 0% ON / 100% OFF**

|      | isTwitterSharingEnabled <br/> 0% ON / 100% OFF | isFacebookSharingEnabled <br/> 0% ON / 100% OFF |
| ---- | ---------------------------------------------- | ----------------------------------------------- |
| Jane | 8 >= 0 <br/>-> **OFF**                         | 64 >= 0 <br/>-> **OFF**                         |
| Joe  | 32 >= 0 <br/>-> **OFF**                        | 12 >= 0 <br/>-> **OFF**                         |

2. **Adjustment to 10% ON / 90% OFF**
In this case both feature flags are enabled for only 10% of users.

|      | isTwitterSharingEnabled <br/> 10% ON / 90% OFF | isFacebookSharingEnabled <br/> 10% ON / 90% OFF |
| ---- | ---------------------------------------------- | ----------------------------------------------- |
| Jane | 8 < 10 <br/>-> **ON**                          | 64 >= 10 <br/>-> **OFF**                        |
| Joe  | 32 >= 10 <br/>-> **OFF**                       | 12 >= 10 <br/>-> **OFF**                        |

:::info

> Despite both feature flags being set to 10% ON / 90% OFF, Jane is only enabled for the `isTwitterSharingEnabled` feature flag. Because Jane and Joe have different identifiers, they're assigned different numbers, which determines their eligibility for each feature flag.

Reviewer: ```suggestion
Despite both feature flags being set to 10% ON / 90% OFF, the `isTwitterSharingEnabled` feature flag is only enabled for Jane. Because Jane and Joe have different identifiers, they're assigned different numbers, which determines their eligibility for each feature.
```
